File: ios/BuzzTerminal/Data/SyncCenter.swift
import Foundation
import OSLog
import SwiftUI

/// The app's view of what has been queued, acknowledged and refused.
///
/// `@MainActor` because the UI reads it every frame, which also makes it `Sendable`
/// and so safe to hand to the repository actor.
///
/// Failures are persisted, and that is the whole point of this type existing rather
/// than a couple of properties on `AppModel`. A refused charge is money that went
/// missing; it has to survive the app being force-quit, the phone dying, and the
/// shift changing.
@MainActor
@Observable
final class SyncCenter {

    private static let log = Logger(subsystem: "fest.swingbuzz.BuzzTerminal", category: "sync")
    private static let storageKey = "fest.swingbuzz.failedWrites"

    private(set) var state = SyncState()

    private let defaults: UserDefaults

    init(defaults: UserDefaults = .standard) {
        self.defaults = defaults
        loadFailures()
    }

    // MARK: - Queue

    func enqueued() {
        state.enqueued()
    }

    func acknowledged() {
        state.acknowledged()
    }

    func failed(_ write: FailedWrite) {
        state.failed(write)
        saveFailures()
        // Logged as well as stored: if the phone is later unavailable, a sysdiagnose
        // still carries the record.
        Self.log.error("""
            write refused: \(write.kind.rawValue, privacy: .public) \
            \(write.amount.description, privacy: .public) \
            participant \(write.participantId, privacy: .public) \
            tx \(write.transactionId, privacy: .public) — \
            \(write.reason, privacy: .public)
            """)
    }

    func settle(_ id: UUID) {
        state.settle(id)
        saveFailures()
    }

    func setOffline(_ offline: Bool) {
        guard state.isOffline != offline else { return }
        state.isOffline = offline
        Self.log.notice("connectivity: \(offline ? "offline" : "online", privacy: .public)")
    }

    #if DEBUG
    /// Put the banner into a given state, for screenshots and previews. Does not
    /// touch the persisted failure list — a fake banner must never be mistaken for
    /// a record of real missing money.
    func simulate(offline: Bool, pending: Int = 0) {
        state.isOffline = offline
        for _ in 0..<pending { state.enqueued() }
    }
    #endif

    // MARK: - Persistence

    private func loadFailures() {
        guard let data = defaults.data(forKey: Self.storageKey) else { return }
        do {
            let decoded = try JSONDecoder().decode([FailedWrite].self, from: data)
            state.replaceFailures(decoded)
            let unsettled = decoded.filter { !$0.settled }.count
            if unsettled > 0 {
                Self.log.error("\(unsettled, privacy: .public) unsettled failed write(s) carried over from a previous run")
            }
        } catch {
            // Deliberately not cleared: a decode failure means something is wrong
            // with the format, and throwing away the only record of missing money
            // to tidy up would be the worst possible response.
            Self.log.error("could not decode failed writes: \(error.localizedDescription, privacy: .public)")
        }
    }

    private func saveFailures() {
        do {
            defaults.set(try JSONEncoder().encode(state.failures), forKey: Self.storageKey)
        } catch {
            Self.log.error("could not persist failed writes: \(error.localizedDescription, privacy: .public)")
        }
    }
}

File: android/app/src/main/kotlin/fest/swingbuzz/terminal/data/SyncCenter.kt
package fest.swingbuzz.terminal.data

import android.content.Context
import android.content.SharedPreferences
import android.util.Log
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import fest.swingbuzz.terminal.domain.FailedWrite
import fest.swingbuzz.terminal.domain.Money
import fest.swingbuzz.terminal.domain.SyncState
import java.time.Instant
import java.util.UUID
import org.json.JSONArray
import org.json.JSONObject

/**
 * The app's view of what has been queued, acknowledged and refused.
 *
 * Compose state, so the UI recomposes; created once and shared between the model
 * and the repository, which is what reports into it.
 *
 * Failures are persisted, and that is the whole point of this class existing rather
 * than a couple of properties on `AppModel`. A refused charge is money that went
 * missing; it has to survive the app being force-stopped, the phone dying, and the
 * shift changing.
 *
 * The JSON is written by hand with `org.json`. `:domain` has no serialisation
 * dependency on purpose, and adding one for six fields would cost more than it
 * saves — the round trip has a test.
 */
class SyncCenter(context: Context?) {

    var state by mutableStateOf(SyncState())
        private set

    private val prefs: SharedPreferences? =
        context?.getSharedPreferences(PREFS, Context.MODE_PRIVATE)

    init {
        loadFailures()
    }

    // ── Queue ──────────────────────────────────────────────────────────────

    fun enqueued() {
        state = state.enqueued()
    }

    fun acknowledged() {
        state = state.acknowledged()
    }

    fun failed(write: FailedWrite) {
        state = state.failed(write)
        saveFailures()
        // Logged as well as stored: if the phone is later unavailable, a bug report
        // still carries the record.
        Log.e(
            TAG,
            "write refused: ${write.kind.wire} ${write.amount} " +
                "participant ${write.participantId} tx ${write.transactionId} — ${write.reason}",
        )
    }

    fun settle(id: UUID) {
        state = state.settle(id)
        saveFailures()
    }

    fun setOffline(offline: Boolean) {
        if (state.isOffline == offline) return
        state = state.copy(isOffline = offline)
        Log.i(TAG, "connectivity: " + if (offline) "offline" else "online")
    }

    /**
     * Put the banner into a given state, for previews and the screenshot pass. Does
     * not touch the persisted failure list — a fake banner must never be mistaken
     * for a record of real missing money.
     */
    fun simulate(offline: Boolean, pending: Int = 0) {
        var next = state.copy(isOffline = offline)
        repeat(pending) { next = next.enqueued() }
        state = next
    }

    // ── Persistence ────────────────────────────────────────────────────────

    private fun loadFailures() {
        val raw = prefs?.getString(KEY, null) ?: return
        try {
            val array = JSONArray(raw)
            val decoded = (0 until array.length()).mapNotNull { decode(array.getJSONObject(it)) }
            state = state.withFailures(decoded)
            val unsettled = decoded.count { !it.settled }
            if (unsettled > 0) {
                Log.e(TAG, "$unsettled unsettled failed write(s) carried over from a previous run")
            }
        } catch (e: Exception) {
            // Deliberately not cleared. A decode failure means something is wrong
            // with the format, and throwing away the only record of missing money to
            // tidy up would be the worst possible response.
            Log.e(TAG, "could not decode failed writes: ${e.message}")
        }
    }

    private fun saveFailures() {
        val array = JSONArray()
        state.failures.forEach { array.put(encode(it)) }
        prefs?.edit()?.putString(KEY, array.toString())?.apply()
    }

    private fun encode(write: FailedWrite) = JSONObject().apply {
        put("id", write.id.toString())
        put("transactionId", write.transactionId)
        put("kind", write.kind.wire)
        put("participantId", write.participantId)
        put("participantName", write.participantName)
        put("braceletId", write.braceletId)
        // Cents, never a decimal: a stored amount must not become a Double on the
        // way to disk and back.
        put("amountCents", write.amount.cents)
        put("attemptedAt", write.attemptedAt.toEpochMilli())
        put("terminalId", write.terminalId)
        put("reason", write.reason)
        put("settled", write.settled)
    }

    private fun decode(json: JSONObject): FailedWrite? {
        val kind = FailedWrite.Kind.fromWire(json.optString("kind")) ?: return null
        return FailedWrite(
            id = runCatching { UUID.fromString(json.getString("id")) }.getOrElse { UUID.randomUUID() },
            transactionId = json.optString("transactionId"),
            kind = kind,
            participantId = json.optString("participantId"),
            participantName = json.optString("participantName"),
            braceletId = json.optString("braceletId"),
            amount = Money(json.optInt("amountCents")),
            attemptedAt = Instant.ofEpochMilli(json.optLong("attemptedAt")),
            terminalId = json.optString("terminalId"),
            reason = json.optString("reason"),
            settled = json.optBoolean("settled"),
        )
    }

    private companion object {
        const val TAG = "SwingBuzz.sync"
        const val PREFS = "fest.swingbuzz.sync"
        const val KEY = "failedWrites"
    }
}
